File: backend/warehouse_service/app/routers/warehouse.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.orm import Session
from typing import List
from datetime import datetime
import logging

from app.database import get_db
from app.models import Location, WarehouseOperation, TempStorageItem
from app.schemas import (
    LocationResponse, LocationStatsResponse, WarehouseOperationCreate,
    WarehouseOperationResponse, TempStorageItemResponse
)
from app.warehouse_service import WarehouseService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/warehouse", tags=["warehouse"])


@router.get("/locations/stats", response_model=List[LocationStatsResponse])
async def get_locations_stats(db: Session = Depends(get_db)):
    """Получить статистику по всем локациям"""
    stats = await WarehouseService.get_location_stats(db)
    return stats


@router.get("/locations", response_model=List[LocationResponse])
async def get_locations(db: Session = Depends(get_db)):
    """Получить все локации"""
    locations = db.query(Location).all()
    return locations


@router.get("/locations/{location_id}", response_model=LocationResponse)
async def get_location(location_id: int, db: Session = Depends(get_db)):
    """Получить локацию по ID"""
    location = db.query(Location).filter(Location.id == location_id).first()
    if not location:
        raise HTTPException(status_code=404, detail="Локация не найдена")
    return location


@router.post("/operations", response_model=WarehouseOperationResponse, status_code=201)
async def create_operation(
    operation: WarehouseOperationCreate,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db)
):
    """Создать операцию перемещения товаров"""
    # Создаем запись операции
    db_operation = WarehouseOperation(
        operation_type=operation.operation_type,
        sku_id=operation.sku_id,
        source_location_id=operation.source_location_id,
        target_location_id=operation.target_location_id,
        quantity_kg=0,  # Будет рассчитано при обработке
        status="pending"
    )
    
    # Получаем название товара если указан ID
    if operation.sku_id:
        try:
            from app.catalog_client import catalog_client
            sku_info = await catalog_client.get_sku(operation.sku_id)
            if sku_info:
                db_operation.sku_name = sku_info.get('name', 'Unknown')
        except Exception as e:
            logger.warning(f"Не удалось получить информацию о товаре {operation.sku_id}: {e}")
            db_operation.sku_name = f"SKU_{operation.sku_id}"
    
    db.add(db_operation)
    db.commit()
    db.refresh(db_operation)
    
    # Обрабатываем операцию в фоне
    background_tasks.add_task(process_operation_background, db_operation.id)
    
    return db_operation


async def process_operation_background(operation_id: int):
    """Фоновая обработка операции"""
    from app.database import SessionLocal
    db = SessionLocal()
    try:
        operation = db.query(WarehouseOperation).filter(WarehouseOperation.id == operation_id).first()
        if not operation:
            logger.error(f"Операция {operation_id} не найдена")
            return
        
        success, error_message = await WarehouseService.process_operation(db, operation)
        
        if success:
            operation.status = "completed"
            operation.completed_at = datetime.now()
            if error_message:
                operation.error_message = error_message  # Предупреждения о недостатке места
        else:
            operation.status = "failed"
            operation.error_message = error_message
        
        db.commit()
        logger.info(f"Операция {operation_id} обработана: {operation.status}")
    except Exception as e:
        logger.error(f"Ошибка при обработке операции {operation_id}: {e}")
        try:
            operation = db.query(WarehouseOperation).filter(WarehouseOperation.id == operation_id).first()
            if operation:
                operation.status = "failed"
                operation.error_message = str(e)
                db.commit()
        except:
            pass
    finally:
        db.close()


@router.get("/operations", response_model=List[WarehouseOperationResponse])
async def get_operations(db: Session = Depends(get_db)):
    """Получить все операции"""
    operations = db.query(WarehouseOperation).order_by(WarehouseOperation.created_at.desc()).all()
    return operations


@router.get("/operations/{operation_id}", response_model=WarehouseOperationResponse)
async def get_operation(operation_id: int, db: Session = Depends(get_db)):
    """Получить операцию по ID"""
    operation = db.query(WarehouseOperation).filter(WarehouseOperation.id == operation_id).first()
    if not operation:
        raise HTTPException(status_code=404, detail="Операция не найдена")
    return operation


@router.get("/temp-storage", response_model=List[TempStorageItemResponse])
async def get_temp_storage_items(db: Session = Depends(get_db)):
    """Получить все элементы временного хранилища"""
    items = db.query(TempStorageItem).filter(
        TempStorageItem.moved_to_storage_at.is_(None)
    ).order_by(TempStorageItem.created_at).all()
    return items


@router.post("/temp-storage/process")
async def process_temp_storage(background_tasks: BackgroundTasks, db: Session = Depends(get_db)):
    """Обработать перемещение товаров из временного хранилища в основное"""
    background_tasks.add_task(WarehouseService.process_temp_storage_migration, db)
    return {"message": "Обработка временного хранилища запущена"}

File: frontend/src/api/warehouse.ts
import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

export type LocationType = 'storage' | 'warehouse' | 'temp_storage';

export type OperationType = 
  | 'receipt' 
  | 'shipment' 
  | 'transfer'
  | 'global_distribution_all'
  | 'global_distribution_sku'
  | 'replenishment_all'
  | 'replenishment_sku'
  | 'placement_all'
  | 'placement_sku';

export interface Location {
  id: number;
  name: string;
  type: LocationType;
  max_capacity_kg: number;
  current_capacity_kg: number;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface LocationStats {
  id: number;
  name: string;
  type: LocationType;
  max_capacity_kg: number;
  current_capacity_kg: number;
  usage_percent: number;
  description?: string;
}

export interface WarehouseOperation {
  id: number;
  operation_type: OperationType;
  sku_id?: number;
  sku_name?: string;
  source_location_id?: number;
  target_location_id?: number;
  quantity_kg: number;
  status: 'pending' | 'completed' | 'failed';
  error_message?: string;
  created_at: string;
  completed_at?: string;
}

export interface WarehouseOperationCreate {
  operation_type: OperationType;
  sku_id?: number;
  source_location_id?: number;
  target_location_id?: number;
}

export interface TempStorageItem {
  id: number;
  sku_id: number;
  sku_name: string;
  quantity_kg: number;
  source_operation_id?: number;
  created_at: string;
  moved_to_storage_at?: string;
}

// Функция для получения заголовков с ролью пользователя
const getHeaders = (): { 'X-User-Role': string } => {
  const userStr = localStorage.getItem('user');
  if (userStr) {
    try {
      const user = JSON.parse(userStr);
      return { 'X-User-Role': user.role || 'viewer' };
    } catch (e) {
      return { 'X-User-Role': 'viewer' };
    }
  }
  return { 'X-User-Role': 'viewer' };
};

const warehouseApi = axios.create({
  baseURL: `${API_URL}/warehouse`,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Добавляем interceptor для добавления заголовка с ролью
warehouseApi.interceptors.request.use((config) => {
  const headers = getHeaders();
  if (config.headers) {
    Object.assign(config.headers, headers);
  }
  return config;
});

export const warehouseService = {
  // Locations
  getLocations: async (): Promise<Location[]> => {
    const response = await warehouseApi.get<Location[]>('/locations');
    return response.data;
  },

  getLocation: async (locationId: number): Promise<Location> => {
    const response = await warehouseApi.get<Location>(`/locations/${locationId}`);
    return response.data;
  },

  getLocationsStats: async (): Promise<LocationStats[]> => {
    const response = await warehouseApi.get<LocationStats[]>('/locations/stats');
    return response.data;
  },

  // Operations
  getOperations: async (): Promise<WarehouseOperation[]> => {
    const response = await warehouseApi.get<WarehouseOperation[]>('/operations');
    return response.data;
  },

  getOperation: async (operationId: number): Promise<WarehouseOperation> => {
    const response = await warehouseApi.get<WarehouseOperation>(`/operations/${operationId}`);
    return response.data;
  },

  createOperation: async (operation: WarehouseOperationCreate): Promise<WarehouseOperation> => {
    const response = await warehouseApi.post<WarehouseOperation>('/operations', operation);
    return response.data;
  },

  // Temp Storage
  getTempStorageItems: async (): Promise<TempStorageItem[]> => {
    const response = await warehouseApi.get<TempStorageItem[]>('/temp-storage');
    return response.data;
  },

  processTempStorage: async (): Promise<{ message: string }> => {
    const response = await warehouseApi.post<{ message: string }>('/temp-storage/process');
    return response.data;
  },
};

